pkg/diff/diff.go

```diff
@@ -0,0 +1,245 @@
+	ChangeTypeKey = "change_type"
+	// ExcludeString defines the string that specifies a path exclusion.
+	ExcludeString = "exclude"
+	// IncludeString defines the string that specifies a path inclusion.
+	// Note: An included path that sits under an excluded path should be included.
+	// For example, if / is excluded, and /files is included, /files should  be added to the archive.
+	IncludeString = "include"
+	// RebaseString defines the path that was stripped when copying from a volume vmdk.
+	// This is necessary as the volume may be mounted deep in the container filesystem,
+	// but files on the volume are seen as living in the filesystem root when being diffed.
+	RebaseString = "rebase"
+)
+
+// TODO(jzt): move this to archive package?
+// PathSpec describes rules for handling specified paths during archival
+type PathSpec struct {
+	Inclusions map[string]struct{}
+	Exclusions map[string]struct{}
+	RebasePath string
+}
+
+// for sort.Sort
+type changesByPath []archive.Change
+
+func (c changesByPath) Less(i, j int) bool { return c[i].Path < c[j].Path }
+func (c changesByPath) Len() int           { return len(c) }
+func (c changesByPath) Swap(i, j int)      { c[j], c[i] = c[i], c[j] }
+
+// Diff produces a tar archive containing the differences between two filesystems
+func Diff(op trace.Operation, newDir, oldDir string, pathspec map[string]string, data bool) (io.ReadCloser, error) {
+
+	changes, err := archive.ChangesDirs(newDir, oldDir)
+	if err != nil {
+		return nil, err
+	}
+
+	sort.Sort(changesByPath(changes))
+
+	spec := processPathSpec(op, pathspec)
+
+	// prune excluded paths from change set
+	for i, change := range changes {
+		if skip(op, change.Path, spec) {
+			changes = append(changes[:i], changes[i+1:]...)
+		}
+	}
+
+	return Tar(op, newDir, changes, spec, data)
+}
+
+func Tar(op trace.Operation, dir string, changes []archive.Change, spec *PathSpec, data bool) (io.ReadCloser, error) {
+	var (
+		err error
+		hdr *tar.Header
+	)
+
+	r, w := io.Pipe()
+	go func() {
+		tw := tar.NewWriter(w)
+
+		for _, change := range changes {
+
+			hdr, err = createHeader(op, dir, change, spec)
+			if err != nil {
+				op.Errorf("Error getting header for change: %s", err.Error())
+				return
+			}
+
+			var f *os.File
+
+			if !data {
+				hdr.Size = 0
+				// include the change type in extended attributes
+				hdr.Xattrs = make(map[string]string)
+				hdr.Xattrs[ChangeTypeKey] = change.Kind.String()
+			}
+
+			_ = tw.WriteHeader(hdr)
+
+			p := path.Join(dir, change.Path)
+			if hdr.Typeflag == tar.TypeReg && hdr.Size != 0 {
+				f, err = os.Open(p)
+				if err != nil {
+					if os.IsPermission(err) {
+						err = nil
+						return
+					}
+					return
+				}
+				if f != nil {
+					_, err = io.Copy(tw, f)
+					_ = f.Close()
+				}
+
+			}
+
+		}
+
+		if err = tw.Close(); err != nil {
+			op.Errorf("Error closing tar writer: %s", err.Error())
+		}
+		if err = w.Close(); err != nil {
```

Comment:
I could be missing something, but we'd be done here - how would the EOF would be considered premature? I should move them up into a `defer` call though, since a `return` caused by an error would skip this closing step.